lora: sync and async transmission

JordanYates opened the pull request:
Change the behavior of `lora_send` to block until transmission has completed.
Add a new function, `lora_send_async` that has the old, non-blocking behavior.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
doc: add LoRa behaviour change to release notes

Document the change in behaviour for `lora_send` and how to get back
the original behaviour.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

andysan commented on drivers/lora/sx12xx_common.c:
```diff
@@ -73,6 +76,15 @@ static void sx12xx_ev_tx_done(void)
 	}
+
+#ifdef CONFIG_LORA_ASYNC
+	if (dev_data.tx_signal) {
```
There seems to be a race in this code since you are releasing the modem before reading the signal. You'll need to keep a local copy of the signal and clear it the dev_data copy before releasing the modem.

JordanYates replied:
Yes. The release does need to stay after the sleep so that the signal can setup another transmission.

JordanYates commented on drivers/lora/sx12xx_common.c:
```diff
@@ -88,6 +100,9 @@ int sx12xx_lora_send(const struct device *dev, uint8_t *data,
 	}
 
+	/* Clear any gives from async transmissions */
+	k_sem_take(&dev_data.tx_rx_done, K_NO_WAIT);
```
Same thing with the other PR, the receive timeout can collide with a packet actually being received, resulting in the semaphore give hanging around until the next call. Reset probably makes this clearer I agree.

Commit:
lora: `lora_send` blocks until completion

Change the behaviour of `lora_send` to block until the transmission
completes. The current asynchronous behaviour is exposed through the
new function `lora_send_async`. This naming convention brings LoRa in
line with other asynchronous subsystems.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
lora: detect failed transmissions

Detect failed transmissions using `lora_send` by only waiting for some
multiple of the actual on-air time before giving up.

We use this instead of the inbuilt TxTimeout functionality because the
value of this timeout is set by `lora_configure`, and therefore doesn't
change with different packet lengths. This is a limitation of the
underlying `RadioSend` function, not the Zephyr driver.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JordanYates commented:
@Mani-Sadhasivam @mniestroj looking for reviews on this so I can progress #36547

Mani-Sadhasivam commented on include/drivers/lora.h:
```diff
@@ -132,6 +143,27 @@ static inline int lora_send(const struct device *dev,
 
+/**
+ * @brief Asynchronously send data over LoRa
+ *
```
Please add a note as like `lora_send` just to align.

JordanYates replied:
done

Mani-Sadhasivam commented on drivers/lora/sx12xx_common.c:
```diff
@@ -88,8 +92,28 @@ int sx12xx_lora_send(const struct device *dev, uint8_t *data,
-	/* Wait for transmission to complete */
-	return k_sem_take(&dev_data.tx_rx_done, K_FOREVER);
+	/* Calculate expected airtime of the packet */
+	air_time = Radio.TimeOnAir(MODEM_LORA,
```
+1. Since the API relies on it explicitly, let's make sure the config is set.

JordanYates replied:
Done

Mani-Sadhasivam commented:
@mniestroj @andysan Please take a look!

mniestroj reviewed (approved):
LGTM

Mani-Sadhasivam commented:
@cfriedt Please merge!
